### ISSUE TITLE
Cross-Segment label offset access bug in 64-bit binary assembly

### ISSUE THREAD

**Issue by ybuzoku** [Terraspace/UASM#162]
Hi there, not sure if this is actually a bug or I am missing something but nevertheless, I've come across what appears to be a very weird bug. This issue manifested itself in JWASM, and again in UASM so I believe if it is a bug, it was inherited from JWASM. It appears that when loading the effective addresses of a label in a different segment to a general purpose register, the assembler will load a phoney addresses instead of the actual offset of that label in the segment, relative to the segment origin. The phoney address is generally a few bytes away from the target address. The example provided below exactly parallels my usecase and the code in which the bug manifests itself.

Given the following code structure, which is being assembled as a flat, 64 bit binary file:

```
.x64p
segment1 SEGMENT BYTE USE64
    ASSUME ds:FLAT, es:FLAT
    ORG 0
...code...code...
    mov rbx, OFFSET myProc    ;The offending line, the wrong value is returned
    add rdx, rbx    ;rdx contains the base address for segment2
    call rdx
...code...code...
segment1 ENDS

segment2 SEGMENT BYTE USE64
    ASSUME ds:FLAT, es:FLAT
    ORG 0
data dSeg <> ;dSeg is an uninitialised struc, a static data area within segment2
    ORG $
...code...code...
myProc PROC
myProc ENDP
segment2 ENDS
```
The idea is that segment1 relocates segment2 to a location given by the program loader at loadtime and then calls a number of functions in segment2 in its relocated location. segment2 uses the data struc as local variable space, and the code in segment2 is relocated in such a way to make space for this data area (I am doing it this way as placing the actual fields of the structure directly in segment2 makes uasm initialise them; this too being undesirable but "livable with" behaviour). 
The issue is that the line i've labelled as "the offending line", seems to load the wrong offset for myProc into rbx, often missing myProc by a few bytes, but I haven't been able to find any pattern to the number of bytes by which it misses myProc. Replacing `mov rbx, OFFSET myProc` with `lea rbx, myProc` does not improve the situation; in both cases the value loaded into rbx does not agree with the value in the listing file for the offset of myProc.

**Comment by john-terraspace** [comment]
Starting looking into this.
The first point to make is that the definition of the dSeg structure must NOT use ?
and the declaration (as you have) must just be <> and not <?> to ensure that the assembler actually generates the bytes in the output BIN file. Using ? in either place gives a correct listing but the BIN file output is treating the space for the struct as BSS (which clearly isn't useful in a bin). I'm not inclined to fix that right now as it might have impacts on other projects, and can be worked around by just ensuring no use of ?.
With that in place, a lea rbx, myProc is producing the correct displacement (assuming segments are contiguous).

`

0:  33 c0                   xor    eax,eax
2:  48 bb 0f 00 00 00 00    movabs rbx,0xf          ; This is not correct
9:  00 00 00
c:  48 8d 1d 0c 00 00 00    lea    rbx,[rip+0xc]        # 0x1f      ; This is correct
13: 48 03 d3                add    rdx,rbx
16: ff d2                   call   rdx
18: 33 c0                   xor    eax,eax
1a: 00 00                   add    BYTE PTR [rax],al
1c: 00 00                   add    BYTE PTR [rax],al
1e: 90                      nop
1f: c3                      ret

`

OFFSET is definitely wrong. I've checked the actual symbol table entry for myProc, and it reflects the correct offset of 5.

**Comment by john-terraspace** [comment]
Found an issue, even with BIN output, Fixups must still be applied, and for BIN files there was no special case handling. OFFSETS were still calculated relative using the segment start offset in the file. I've kept it that way for intra-segment offsets, but for inter segment fixups it will now just use the offset relative to the base of the segment the symbol is in. I've only applied this fix for 64bit BIN to limit blast radius.

Output now looks like:
`

0:  33 c0                   xor    eax,eax
2:  48 bb 05 00 00 00 00    movabs rbx,0x5
9:  00 00 00
c:  48 8d 1d 0c 00 00 00    lea    rbx,[rip+0xc]        # 0x1f
13: 48 03 d3                add    rdx,rbx
16: ff d2                   call   rdx
18: 33 c0                   xor    eax,eax
1a: 00 00                   add    BYTE PTR [rax],al
1c: 00 00                   add    BYTE PTR [rax],al
1e: 90                      nop
1f: c3                      ret

`
